src/backend/gl/src/device.rs

```diff
@@ -519,133 +513,137 @@ impl Device {
-                    Ok(texture_mapping) => {
-                        Self::reflect_shader(
-                            module,
-                            &options.entry_point,
-                            texture_mapping,
-                            sampler_map,
-                            name_binding_map,
-                            layout,
-                        );
-                        let source = String::from_utf8(output).unwrap();
-                        debug!("Naga generated shader:\n{}", source);
-                        self.create_shader_module_raw(&source, stage).unwrap()
-                    }
-                    Err(e) => {
-                        warn!("Naga: {}", e);
-                        let fallback = pso::EntryPoint {
-                            module: &n::ShaderModule::Spirv(spirv.to_vec()),
-                            ..point.clone()
-                        };
-                        self.compile_shader(&fallback, stage, layout, sampler_map, name_binding_map)
-                    }
+    fn compile_shader(
+        &self,
+        ep: &pso::EntryPoint<B>,
+        stage: naga::ShaderStage,
+        mut context: CompilationContext,
+    ) -> Result<n::Shader, d::ShaderError> {
+        let naga_options = naga::back::glsl::Options {
+            version: {
+                use naga::back::glsl::Version;
+                let sl = &self.share.info.shading_language;
+                let value = (sl.major * 100 + sl.minor * 10) as u16;
+                if sl.is_embedded {
+                    Version::Embedded(value)
+                } else {
+                    Version::Desktop(value)
                 }
+            },
+            entry_point: (stage, ep.entry.to_string()),
+        };
+
+        let mut result = Err(d::ShaderError::CompilationFailed(String::new()));
+        if ep.module.prefer_naga {
+            if let Some(ref module) = ep.module.naga {
+                result = Self::compile_shader_library_naga(
+                    &self.share.context,
+                    module,
+                    &naga_options,
+                    context.reborrow(),
+                );
             }
         }
+        #[cfg(feature = "cross")]
+        if result.is_err() {
+            let mut ast = self.parse_spirv_cross(&ep.module.spv).unwrap();
+            auxil::spirv_cross_specialize_ast(&mut ast, &ep.specialization).unwrap();
+            self.remap_bindings(&mut ast, context.layout, context.name_binding_map);
+            self.combine_separate_images_and_samplers(&mut ast, context.reborrow());
+            self.set_push_const_layout(&mut ast).unwrap();
+
+            let glsl = self
+                .translate_spirv_cross(&mut ast, stage, ep.entry)
+                .unwrap();
+            debug!("SPIRV-Cross generated shader:\n{}", glsl);
+            result = Self::create_shader_module_raw(&self.share.context, &glsl, stage);
+        }
+        if !ep.module.prefer_naga {
```

Comment:
@kvark Awesome work! Thank you!

Maybe you mean:
```rust
if ep.module.prefer_naga {
```

Otherwise, we still get naga errors with the enabled `cross` feature.

Reply:
The point here is to try Naga in case it was not preferred originally. I was indeed missing a check for the result here, thank you for spotting!